### PR TITLE
[Dist] Allow reading and writing single-column vector Parquet files.

### PR DESCRIPTION
**Pull request by thvasilo**
## Description
[Spark ML feature processing](https://spark.apache.org/docs/latest/ml-features) will commonly produce DataFrames, and by extension Parquet files, that contain columns whose elements are vectors instead of single values.

Example:

```python
>>> table.to_pandas()
      vector
0  [1, 2, 3]
1  [4, 5, 6]
>>> table.shape
(2, 1)
```

This PR adds the capability to the Parquet reader to read such files by transforming them to regular Numpy matrices, converting each vector to one row of the matrix.

@classicsong @Rhett-Ying  requested for review

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: tests/tools/utils.py [hunk 1]
```diff
@@ -9,15 +9,20 @@
...
     offset = 0
 
     for j, n in enumerate(chunk_sizes):
         path = os.path.abspath(path_fmt % j)
         arr_chunk = arr[offset: offset + n]
+        shape = arr_chunk.shape
         logging.info("Chunking %d-%d" % (offset, offset + n))
-        array_readwriter.get_array_parser(**fmt_meta).write(path, arr_chunk)
+        # If requested we write multi-column arrays as single-column vector files
+        if vector_rows and len(shape) > 1 and shape[1] > 1:
```
Comment (thvasilo) [inline tests/tools/utils.py:22]: This check can be pushed down to the parquet writer, not sure what is preferred here. 

To guarantee that any Parquet file written using the parser can be read using the parser we might need to do that.

File: tools/distpartitioning/array_readwriter/parquet.py [hunk 2]
```diff
@@ -16,26 +17,44 @@ def read(self, path):
...
             logging.warning(
                 "Shape information not found in the metadata, read the data as "
                 "a 2 dim array."
             )
+        logging.info("Done reading from %s" % path)
         shape = tuple(eval(shape.decode())) if shape else arr.shape
         return arr.reshape(shape)
 
-    def write(self, path, array):
+    def write(self, path, array, vector_rows=False):
```
Comment (thvasilo) [inline tools/distpartitioning/array_readwriter/parquet.py:45]: The change mostly relates to the reader, but in order to write the unit tests in a non-invasive way I added some modifications to the writer as well.

An alternative is to write a specific Parquet writer just for the tests and avoid making this modification to the writer.

**Comment by Rhett-Ying** [comment]
@dgl-bot

File: tests/tools/utils.py (hunk 1 above)
Comment (Rhett-Ying) [reply tests/tools/utils.py:22]: writing a single-column is only available for `parquet`, right? we'd better add a format check here to capture any potential use `vector_rows=True` with `numpy`.

File: tools/distpartitioning/array_readwriter/parquet.py (hunk 2 above)
Comment (Rhett-Ying) [reply tools/distpartitioning/array_readwriter/parquet.py:45]: I think it's ok to add new arg here

**Review by Rhett-Ying — APPROVED** [review]
LGTM

File: tests/tools/utils.py (hunk 1 above)
Comment (thvasilo) [reply tests/tools/utils.py:22]: It is actually possible with Numpy too, using a [structured array](https://numpy.org/doc/stable/user/basics.rec.html), but not sure if/why one would produce such a file if they already have a regular 2D `ndarray`.

My question is more about where these checks should exist, inside `ParquetArrayParser` or only at the tests?

Comment (Rhett-Ying) [reply tests/tools/utils.py:22]: only at tests should be ok as `.write()` is called here for now.

Comment (thvasilo) [reply tests/tools/utils.py]: OK the new commit should make the writer only use vector_rows when we're writing to Parquet.

**Comment by Rhett-Ying** [comment]
@dgl-bot

**Comment by Rhett-Ying** [comment]
@dgl-bot

**Pull request merged**